File: Tests/CSharp/Roslyn/Syntax/BaseTypeDeclarationExtensions.fs
// The MIT License (MIT)
// open SafetySharp.Modeling
// Copyright (c) 2014-2015, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace Roslyn.Syntax.BaseTypeDeclarationExtensions

open System
open System.Linq
open System.Runtime.CompilerServices
open NUnit.Framework
open Microsoft.CodeAnalysis
open Microsoft.CodeAnalysis.CSharp.Syntax
open SafetySharp.Modeling.CompilerServices
open SafetySharp.CSharp.Roslyn.Syntax
open SafetySharp.CSharp.Roslyn.Symbols

[<TestFixture>]
module ``GetTypeSymbol method`` =
    let checkTypeSymbol csharpCode typeName =
        let compilation = TestCompilation csharpCode
        let declaration = compilation.FindTypeDeclaration typeName
        let symbol = compilation.FindTypeSymbol typeName

        declaration.GetTypeSymbol compilation.SemanticModel =? symbol

    [<Test>]
    let ``throws when type declaration is null`` () =
        let compilation = TestCompilation ""
        raisesArgumentNullException "typeDeclaration" (fun () -> (null : BaseTypeDeclarationSyntax).GetTypeSymbol compilation.SemanticModel |> ignore)

    [<Test>]
    let ``throws when semantic model is null`` () =
        let compilation = TestCompilation "class X {}"
        let declaration = compilation.FindClassDeclaration "X"
        raisesArgumentNullException "semanticModel" (fun () -> declaration.GetTypeSymbol null |> ignore)

    [<Test>]
    let ``returns correct class symbol`` () =
        checkTypeSymbol "class X {} class Y {}" "X"
        checkTypeSymbol "class X {} class Y {}" "Y"

    [<Test>]
    let ``returns correct interface symbol`` () =
        checkTypeSymbol "interface X {} class Y {}" "X"
        checkTypeSymbol "class X {} interface Y {}" "Y"
        checkTypeSymbol "interface X {} interface Y {}" "Y"

    [<Test>]
    let ``returns correct enum symbol`` () =
        checkTypeSymbol "enum X {} class Y {}" "X"
        checkTypeSymbol "class X {} enum Y {}" "Y"
        checkTypeSymbol "enum X {} enum Y {}" "Y"

    [<Test>]
    let ``returns correct struct symbol`` () =
        checkTypeSymbol "struct X {} class Y {}" "X"
        checkTypeSymbol "class X {} struct Y {}" "Y"
        checkTypeSymbol "struct X {} struct Y {}" "Y"

[<TestFixture>]
module ``IsDerivedFrom method`` =
    let isDerivedFrom csharpCode baseName =
        let compilation = TestCompilation csharpCode

        let derivedDeclaration = compilation.FindTypeDeclaration "X"
        let baseDeclaration = compilation.FindTypeSymbol baseName

        derivedDeclaration.IsDerivedFrom (compilation.SemanticModel, baseDeclaration)

    [<Test>]
    let ``throws when type declaration is null`` () =
        let compilation = TestCompilation "class X {}"
        let symbol = compilation.FindClassSymbol "X"
        raisesArgumentNullException "typeDeclaration" 
            (fun () -> (null : BaseTypeDeclarationSyntax).IsDerivedFrom (compilation.SemanticModel, symbol) |> ignore)

    [<Test>]
    let ``throws when semantic model is null`` () =
        let compilation = TestCompilation "class X {}"
        let declaration = compilation.FindClassDeclaration "X"
        let symbol = compilation.FindClassSymbol "X"
        raisesArgumentNullException "semanticModel" (fun () -> declaration.IsDerivedFrom (null, symbol) |> ignore)

    [<Test>]
    let ``throws when base type symbol is null`` () =
        let compilation = TestCompilation "class X {}"
        let declaration = compilation.FindClassDeclaration "X"
        raisesArgumentNullException "baseType" (fun () -> declaration.IsDerivedFrom (compilation.SemanticModel, null) |> ignore)

    [<Test>]
    let ``throws when base type is null`` () =
        let compilation = TestCompilation ""
        let symbol = compilation.CSharpCompilation.GetTypeSymbol<obj> ()
        raisesArgumentNullException "baseType" (fun () -> symbol.IsDerivedFrom null |> ignore)

    [<Test>]
    let ``returns false for self checks`` () =
        isDerivedFrom "interface X {}" "X" =? false
        isDerivedFrom "class X {}" "X" =? false

    [<Test>]
    let ``returns false when class does not derive from given base class`` () =
        isDerivedFrom "class Q {} class X {}" "Q" =? false
        isDerivedFrom "class Q {} class Y {} class X : Y {}" "Q" =? false
        isDerivedFrom "class Q {} class Z {} class Y : Z {} class X : Y {}" "Q" =? false

    [<Test>]
    let ``returns false when interface does not derive from given base interface`` () =
        isDerivedFrom "interface Q {} interface X {}" "Q" =? false
        isDerivedFrom "interface Q {} interface Y {} interface X : Y {}" "Q" =? false
        isDerivedFrom "interface Q {} interface Z {} interface Y : Z {} interface X : Y {}" "Q" =? false

    [<Test>]
    let ``returns true when class derives from given base class`` () =
        isDerivedFrom "class Y {} class X : Y {}" "Y" =? true
        isDerivedFrom "class Z {} class Y : Z {} class X : Y {}" "Y" =? true
        isDerivedFrom "class Z {} class Y : Z {} class X : Y {}" "Z" =? true

    [<Test>]
    let ``returns true when class has base class that derives from the given base interface`` () =
        isDerivedFrom "interface Q {} interface Z {} class Y : Z, Q {} class X : Y {}" "Q" =? true
        isDerivedFrom "interface Q {} interface Z {} class Y : Z, Q {} class X : Y {}" "Z" =? true
        isDerivedFrom "interface S {} interface Q {} class Z : Q, S {} class Y : Z, Q {} class X : Y {}" "Q" =? true
        isDerivedFrom "interface S {} interface Q {} class Z : Q, S {} class Y : Z, Q {} class X : Y {}" "S" =? true

    [<Test>]
    let ``returns true when interface derives from given base interface`` () =
        isDerivedFrom "interface Y {} interface X : Y {}" "Y" =? true
        isDerivedFrom "interface Z {} interface Y : Z {} interface X : Y {}" "Y" =? true
        isDerivedFrom "interface Z {} interface Y : Z {} interface X : Y {}" "Z" =? true
        isDerivedFrom "interface Z {} interface Y : Z {} interface X : Y, Z {}" "Y" =? true
        isDerivedFrom "interface Z {} interface Y : Z {} interface X : Y, Z {}" "Z" =? true
        isDerivedFrom "interface Z {} interface Y {} interface X : Y, Z {}" "Y" =? true
        isDerivedFrom "interface Z {} interface Y {} interface X : Y, Z {}" "Z" =? true
        isDerivedFrom "interface Q {} interface Z {} interface Y : Z, Q {} interface X : Y {}" "Q" =? true

[<TestFixture>]
module ``IsDerivedFromComponent methods`` =
    let isDerivedFromComponent csharpCode =
        let compilation = TestCompilation csharpCode
        let derivedDeclaration = compilation.FindTypeDeclaration "X"

        derivedDeclaration.IsDerivedFromComponent compilation.SemanticModel

    [<Test>]
    let ``throws when type declaration is null`` () =
        let compilation = TestCompilation("class X {}")
        raisesArgumentNullException "typeDeclaration" (fun () -> (null : BaseTypeDeclarationSyntax).IsDerivedFromComponent compilation.SemanticModel |> ignore)

    [<Test>]
    let ``throws when semantic model is null`` () =
        let declaration = TestCompilation("class X {}").FindClassDeclaration "X"
        raisesArgumentNullException "semanticModel" (fun () -> declaration.IsDerivedFromComponent (null : SemanticModel) |> ignore)

    [<Test>]
    let ``returns false for class with no base`` () =
        isDerivedFromComponent "class X {}" =? false

    [<Test>]
    let ``returns false for class with non-Component base`` () =
        isDerivedFromComponent "class Y { } class X : Y {}" =? false

    [<Test>]
    let ``returns false for class only implementing IComponent`` () =
        isDerivedFromComponent "class X : IComponent { public void Update() {} }" =? false

    [<Test>]
    let ``returns true for class directly derived from Component`` () =
        isDerivedFromComponent "class X : Component {}" =? true

    [<Test>]
    let ``returns true for class indirectly derived from Component`` () =
        isDerivedFromComponent "class Y : Component {} class X : Y {}" =? true

[<TestFixture>]
module ``ImplementsIComponent methods`` =
    let implementsIComponent csharpCode =
        let compilation = TestCompilation csharpCode
        let derivedDeclaration = compilation.FindTypeDeclaration "X"

        derivedDeclaration.ImplementsIComponent compilation.SemanticModel

    [<Test>]
    let ``throws when type declaration is null`` () =
        let compilation = TestCompilation("class X {}")
        raisesArgumentNullException "typeDeclaration" (fun () -> (null : BaseTypeDeclarationSyntax).ImplementsIComponent compilation.SemanticModel |> ignore)

    [<Test>]
    let ``throws when semantic model is null`` () =
        let symbol = TestCompilation("class X {}").FindTypeSymbol("X")
        raisesArgumentNullException "semanticModel" (fun () -> symbol.ImplementsIComponent (null : SemanticModel) |> ignore)

    [<Test>]
    let ``returns false for class with no base`` () =
        implementsIComponent "class X {}" =? false

    [<Test>]
    let ``returns false for class with non-Component base`` () =
        implementsIComponent "class Y { } class X : Y {}" =? false

    [<Test>]
    let ``returns true for class only implementing IComponent`` () =
        implementsIComponent "class X : IComponent { public void Update() {} }" =? true

    [<Test>]
    let ``returns true for class directly derived from Component`` () =
        implementsIComponent "class X : Component {}" =? true

    [<Test>]
    let ``returns true for class indirectly derived from Component`` () =
        implementsIComponent "class Y : Component {} class X : Y {}" =? true

    [<Test>]
    let ``returns true for class indirectly implementing IComponent`` () =
        implementsIComponent "class Y : IComponent { public void Update() {} } class X : Y {}" =? true

File: Source/CSharp/Roslyn/Syntax/BaseTypeDeclarationExtensions.cs
﻿// The MIT License (MIT)
// 
// Copyright (c) 2014-2015, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace SafetySharp.CSharp.Roslyn.Syntax
{
	using System;
	using JetBrains.Annotations;
	using Microsoft.CodeAnalysis;
	using Microsoft.CodeAnalysis.CSharp;
	using Microsoft.CodeAnalysis.CSharp.Syntax;
	using Modeling;
	using Symbols;
	using Utilities;

	/// <summary>
	///     Provides extension methods for working with <see cref="BaseTypeDeclarationSyntax" /> instances.
	/// </summary>
	public static class BaseTypeDeclarationExtensions
	{
		/// <summary>
		///     Gets the <see cref="INamedTypeSymbol" /> declared by <paramref name="typeDeclaration" /> within the context of the
		///     <paramref name="semanticModel" />.
		/// </summary>
		/// <param name="typeDeclaration">The type declaration the declared symbol should be returned for.</param>
		/// <param name="semanticModel">The semantic model that should be used to determine the declared symbol.</param>
		[Pure, NotNull]
		public static INamedTypeSymbol GetTypeSymbol([NotNull] this BaseTypeDeclarationSyntax typeDeclaration,
													 [NotNull] SemanticModel semanticModel)
		{
			Requires.NotNull(typeDeclaration, () => typeDeclaration);
			Requires.NotNull(semanticModel, () => semanticModel);

			var symbol = semanticModel.GetDeclaredSymbol(typeDeclaration);
			Assert.NotNull(symbol, "Unable to determine type symbol of type declaration '{0}'.", typeDeclaration);

			return symbol;
		}

		/// <summary>
		///     Checks whether <paramref name="typeDeclaration" /> is directly or indirectly derived from the
		///     <paramref name="baseType" />
		///     interface or class within the context of the <paramref name="semanticModel" />.
		/// </summary>
		/// <param name="typeDeclaration">The type declaration that should be checked.</param>
		/// <param name="semanticModel">
		///     The semantic model that should be used to resolve the type symbol of the <paramref name="typeDeclaration" />.
		/// </param>
		/// <param name="baseType">The base type interface or class that <paramref name="typeDeclaration" /> should be derived from.</param>
		[Pure]
		public static bool IsDerivedFrom([NotNull] this BaseTypeDeclarationSyntax typeDeclaration, [NotNull] SemanticModel semanticModel,
										 [NotNull] ITypeSymbol baseType)
		{
			Requires.NotNull(typeDeclaration, () => typeDeclaration);
			Requires.NotNull(semanticModel, () => semanticModel);
			Requires.NotNull(baseType, () => baseType);

			return typeDeclaration.GetTypeSymbol(semanticModel).IsDerivedFrom(baseType);
		}

		/// <summary>
		///     Checks whether <paramref name="typeDeclaration" /> is directly or indirectly derived from the <see cref="Component" />
		///     class within the context of the <paramref name="semanticModel" />.
		/// </summary>
		/// <param name="typeDeclaration">The type declaration that should be checked.</param>
		/// <param name="semanticModel">
		///     The semantic model that should be used to resolve the type symbol for the <see cref="Component" /> class.
		/// </param>
		[Pure]
		public static bool IsDerivedFromComponent([NotNull] this BaseTypeDeclarationSyntax typeDeclaration, [NotNull] SemanticModel semanticModel)
		{
			Requires.NotNull(typeDeclaration, () => typeDeclaration);
			Requires.NotNull(semanticModel, () => semanticModel);

			return typeDeclaration.IsDerivedFrom(semanticModel, semanticModel.GetComponentClassSymbol());
		}

		/// <summary>
		///     Checks whether <paramref name="typeDeclaration" /> directly or indirectly implements the <see cref="IComponent" />
		///     interface within the context of the <paramref name="semanticModel" />.
		/// </summary>
		/// <param name="typeDeclaration">The type declaration that should be checked.</param>
		/// <param name="semanticModel">
		///     The semantic model that should be used to resolve the type symbol for the <see cref="IComponent" /> interface.
		/// </param>
		[Pure]
		public static bool ImplementsIComponent([NotNull] this BaseTypeDeclarationSyntax typeDeclaration, [NotNull] SemanticModel semanticModel)
		{
			Requires.NotNull(typeDeclaration, () => typeDeclaration);
			Requires.NotNull(semanticModel, () => semanticModel);

			return typeDeclaration.IsDerivedFrom(semanticModel, semanticModel.GetComponentInterfaceSymbol());
		}
	}
}